File: examples/mv_ipc_client.rs
use std::env;
use std::io::{Read, Write};
use std::os::unix::net::UnixStream;
use std::process;

const DEFAULT_SOCKET_PATH: &str = "/tmp/fro-mv.sock";

fn usage(program: &str) {
    eprintln!("USAGE: {program} <source> <target>");
}

fn main() {
    let mut args = env::args_os();
    let program = args
        .next()
        .unwrap_or_else(|| "mv_ipc_client_rust".into())
        .to_string_lossy()
        .into_owned();
    let Some(source) = args.next() else {
        usage(&program);
        process::exit(1);
    };
    let Some(target) = args.next() else {
        usage(&program);
        process::exit(1);
    };
    if args.next().is_some() {
        usage(&program);
        process::exit(1);
    }

    let mut stream = match UnixStream::connect(DEFAULT_SOCKET_PATH) {
        Ok(stream) => stream,
        Err(err) => {
            eprintln!("{program}: {err}");
            process::exit(1);
        }
    };
    let source_bytes = source.to_string_lossy();
    let target_bytes = target.to_string_lossy();
    if stream.write_all(source_bytes.as_bytes()).is_err()
        || stream.write_all(&[0]).is_err()
        || stream.write_all(target_bytes.as_bytes()).is_err()
        || stream.write_all(&[0]).is_err()
        || stream.shutdown(std::net::Shutdown::Write).is_err()
    {
        process::exit(1);
    }
    let mut status = [1u8; 1];
    if stream.read_exact(&mut status).is_err() {
        process::exit(1);
    }
    process::exit(status[0] as i32);
}

File: examples/mv_ipc_client.c
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>

#define SOCKET_PATH "/tmp/fro-mv.sock"

static int write_all(int fd, const void *buf, size_t len) {
    const char *ptr = (const char *)buf;
    while (len > 0) {
        ssize_t written = write(fd, ptr, len);
        if (written <= 0) {
            return -1;
        }
        ptr += written;
        len -= (size_t)written;
    }
    return 0;
}

int main(int argc, char **argv) {
    struct sockaddr_un addr;
    int fd;
    unsigned char status = 1;

    if (argc != 3) {
        fprintf(stderr, "USAGE: %s <source> <target>\n", argv[0]);
        return 1;
    }

    fd = socket(AF_UNIX, SOCK_STREAM, 0);
    if (fd < 0) {
        return 1;
    }

    memset(&addr, 0, sizeof(addr));
    addr.sun_family = AF_UNIX;
    memcpy(addr.sun_path, SOCKET_PATH, sizeof(SOCKET_PATH));
    if (connect(fd, (struct sockaddr *)&addr, sizeof(sa_family_t) + sizeof(SOCKET_PATH)) != 0) {
        close(fd);
        return 1;
    }

    if (write_all(fd, argv[1], strlen(argv[1])) != 0 ||
        write_all(fd, "\0", 1) != 0 ||
        write_all(fd, argv[2], strlen(argv[2])) != 0 ||
        write_all(fd, "\0", 1) != 0 ||
        shutdown(fd, SHUT_WR) != 0 ||
        read(fd, &status, 1) != 1) {
        close(fd);
        return 1;
    }

    close(fd);
    return status;
}
